File: src/api/packages.ts
import { invoke } from '@tauri-apps/api/core';
import { PackageSource, type Package } from '../types/package';
import { createLogger } from '../utils/logger';

const log = createLogger('api');

export async function searchPackages(
  query: string,
  sources?: PackageSource[]
): Promise<Package[]> {
  log.info(`search("${query}", sources=${sources?.join(',') ?? 'all'})`);
  const result = await invoke<Package[]>('search_packages', {
    query,
    sources: sources?.map((s) => s),
  });
  log.info(`search => ${result.length} results`);
  return result;
}

export async function getFeaturedPackages(limit: number = 20): Promise<Package[]> {
  log.info(`getFeatured(limit=${limit})`);
  const result = await invoke<Package[]>('get_featured_packages', { limit });
  log.info(`getFeatured => ${result.length} apps`);
  return result;
}

export async function getScreenshots(
  appIds: string[]
): Promise<Array<[string, string | null]>> {
  log.info(`getScreenshots(${appIds.length} apps)`);
  const result = await invoke<Array<[string, string | null]>>('get_screenshots', { appIds });
  log.info(`getScreenshots => ${result.filter(([, s]) => s !== null).length} with screenshots`);
  return result;
}

export async function getPackageDetail(
  id: string,
  source: PackageSource
): Promise<Package> {
  log.info(`getDetail("${id}", ${source})`);
  const result = await invoke<Package>('get_package_detail', { id, source });
  log.info(`getDetail => "${result?.name ?? 'null'}"`);
  return result;
}

export async function listInstalled(): Promise<Package[]> {
  log.info('listInstalled()');
  const result = await invoke<Package[]>('list_installed');
  log.info(`listInstalled => ${result.length} packages`);
  return result;
}

export async function listArcwayInstalled(): Promise<Package[]> {
  log.info('listArcwayInstalled()');
  const result = await invoke<Package[]>('list_arcway_installed');
  log.info(`listArcwayInstalled => ${result.length} packages`);
  return result;
}

export async function checkArcwayUpdates(): Promise<
  Array<{
    package_id: string;
    name: string;
    source: PackageSource;
    current_version: string;
    new_version: string;
  }>
> {
  log.info('checkArcwayUpdates()');
  const result = await invoke<Array<{ package_id: string; name: string; source: PackageSource; current_version: string; new_version: string }>>('check_arcway_updates');
  log.info(`checkArcwayUpdates => ${result.length} updates`);
  return result;
}

export async function checkUpdates(): Promise<
  Array<{
    package_id: string;
    name: string;
    source: PackageSource;
    current_version: string;
    new_version: string;
  }>
> {
  log.info('checkUpdates()');
  const result = await invoke<Array<{ package_id: string; name: string; source: PackageSource; current_version: string; new_version: string }>>('check_updates');
  log.info(`checkUpdates => ${result.length} updates`);
  return result;
}

export async function getFlathubDetail(appId: string): Promise<Record<string, unknown>> {
  log.info(`getFlathubDetail("${appId}")`);
  const result = await invoke<Record<string, unknown>>('get_flathub_detail', { appId });
  log.info(`getFlathubDetail => OK`);
  return result;
}

export async function getTrendingPackages(limit: number = 30): Promise<Package[]> {
  log.info(`getTrending(limit=${limit})`);
  const result = await invoke<Package[]>('get_trending_packages', { limit });
  log.info(`getTrending => ${result.length} apps`);
  return result;
}

export async function getRecentlyUpdatedPackages(limit: number = 30): Promise<Package[]> {
  log.info(`getRecentlyUpdated(limit=${limit})`);
  const result = await invoke<Package[]>('get_recently_updated_packages', { limit });
  log.info(`getRecentlyUpdated => ${result.length} apps`);
  return result;
}

export async function getRecentlyAddedPackages(limit: number = 30): Promise<Package[]> {
  log.info(`getRecentlyAdded(limit=${limit})`);
  const result = await invoke<Package[]>('get_recently_added_packages', { limit });
  log.info(`getRecentlyAdded => ${result.length} apps`);
  return result;
}

export async function getCategoryPackages(category: string, limit: number = 30): Promise<Package[]> {
  log.info(`getCategory("${category}", limit=${limit})`);
  const result = await invoke<Package[]>('get_category_packages', { category, limit });
  log.info(`getCategory => ${result.length} apps`);
  return result;
}

File: src-tauri/src/commands/packages.rs
use crate::api;
use crate::cache::Cache;
use crate::errors::AppError;
use crate::models::*;
use crate::package_managers::MultiManager;
use tauri::State;

fn score_relevance(pkg: &Package, query: &str) -> u32 {
    let q = query.to_lowercase();
    let name_lower = pkg.name.to_lowercase();
    let id_lower = pkg.id.to_lowercase();

    if name_lower == q {
        100
    } else if name_lower.starts_with(&q) {
        80
    } else if name_lower.contains(&q) {
        60
    } else if id_lower.contains(&q) {
        40
    } else if pkg.description.to_lowercase().contains(&q) {
        20
    } else if pkg.tags.iter().any(|t| t.to_lowercase().contains(&q)) {
        10
    } else {
        0
    }
}

#[tauri::command]
pub async fn search_packages(
    query: String,
    sources: Option<Vec<PackageSource>>,
    _manager: State<'_, MultiManager>,
) -> Result<Vec<Package>, AppError> {
    let mut all_packages = Vec::new();

    let flathub_future = api::flathub::search(&query);
    let aur_future = api::aur::search(&query);

    let (flathub_result, aur_result) = tokio::join!(flathub_future, aur_future);

    match flathub_result {
        Ok(packages) => all_packages.extend(packages),
        Err(e) => {
            log::warn!("Flathub search failed: {}", e);
        }
    }

    match aur_result {
        Ok(packages) => all_packages.extend(packages),
        Err(e) => {
            log::warn!("AUR search failed: {}", e);
        }
    }

    if let Some(source_list) = sources {
        all_packages.retain(|pkg| source_list.contains(&pkg.source));
    }

    // Deduplicate by name+source, keep first occurrence
    let mut seen = std::collections::HashSet::new();
    all_packages.retain(|pkg| {
        let key = format!("{}:{:?}", pkg.name, pkg.source);
        seen.insert(key)
    });

    // Sort by relevance score descending, then alphabetically
    let query_str = query.clone();
    all_packages.sort_by(|a, b| {
        let sa = score_relevance(a, &query_str);
        let sb = score_relevance(b, &query_str);
        sb.cmp(&sa).then_with(|| a.name.cmp(&b.name))
    });

    Ok(all_packages)
}

#[tauri::command]
pub async fn get_featured_packages(
    limit: Option<usize>,
) -> Result<Vec<Package>, AppError> {
    let limit = limit.unwrap_or(20);
    api::flathub::fetch_featured(limit).await
}

#[tauri::command]
pub async fn get_screenshots(
    app_ids: Vec<String>,
) -> Result<Vec<(String, Option<String>)>, AppError> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(8))
        .build()
        .map_err(|e| AppError::CommandFailed(format!("HTTP client error: {}", e)))?;

    let futures: Vec<_> = app_ids
        .iter()
        .map(|id| {
            let url = format!("https://flathub.org/api/v2/appstream/{}", id);
            let client = client.clone();
            let id = id.clone();
            async move {
                let screenshot = match client.get(&url).timeout(std::time::Duration::from_secs(4)).send().await {
                    Ok(resp) => match resp.text().await {
                        Ok(text) => {
                            let detail: Option<serde_json::Value> = serde_json::from_str(&text).ok();
                            detail.and_then(|d| {
                                d.get("screenshots")?
                                    .as_array()?
                                    .first()?
                                    .get("sizes")?
                                    .as_array()?
                                    .iter()
                                    .filter_map(|s| s.get("src").and_then(|v| v.as_str()).map(String::from))
                                    .max_by_key(|s| s.len())
                            })
                        }
                        Err(_) => None,
                    },
                    Err(_) => None,
                };
                (id, screenshot)
            }
        })
        .collect();

    Ok(futures::future::join_all(futures).await)
}

#[tauri::command]
pub async fn get_package_detail(
    id: String,
    source: PackageSource,
    manager: State<'_, MultiManager>,
) -> Result<Option<Package>, AppError> {
    let name = id.split('/').last().unwrap_or(&id);
    let packages = manager.search_in_sources(name, &[source]).await?;
    Ok(packages.into_iter().find(|p| p.id == id || p.name == name))
}

#[tauri::command]
pub async fn list_installed(
    manager: State<'_, MultiManager>,
    cache: State<'_, Cache>,
) -> Result<Vec<InstalledPackage>, AppError> {
    let installed = manager.list_all_installed().await?;
    if let Err(e) = cache.save_installed(&installed) {
        log::warn!("Failed to update installed cache: {}", e);
    }
    Ok(installed)
}

#[tauri::command]
pub async fn get_flathub_detail(app_id: String) -> Result<serde_json::Value, AppError> {
    api::flathub::fetch_app_detail(&app_id).await
}

#[tauri::command]
pub async fn get_trending_packages(
    limit: Option<usize>,
) -> Result<Vec<Package>, AppError> {
    let limit = limit.unwrap_or(30);
    api::flathub::fetch_trending(limit).await
}

#[tauri::command]
pub async fn get_recently_updated_packages(
    limit: Option<usize>,
) -> Result<Vec<Package>, AppError> {
    let limit = limit.unwrap_or(30);
    api::flathub::fetch_recently_updated(limit).await
}

#[tauri::command]
pub async fn get_recently_added_packages(
    limit: Option<usize>,
) -> Result<Vec<Package>, AppError> {
    let limit = limit.unwrap_or(30);
    api::flathub::fetch_recently_added(limit).await
}

#[tauri::command]
pub async fn get_category_packages(
    category: String,
    limit: Option<usize>,
) -> Result<Vec<Package>, AppError> {
    let limit = limit.unwrap_or(30);
    api::flathub::fetch_by_category(&category, limit).await
}

#[tauri::command]
pub async fn list_arcway_installed(
    cache: State<'_, Cache>,
) -> Result<Vec<Package>, AppError> {
    let packages = cache.get_arcway_installed_packages()?;
    log::info!("list_arcway_installed => {} packages", packages.len());
    Ok(packages)
}
